be/src/exec/pipeline/scan/olap_scan_operator.cpp

```diff
@@ -31,6 +31,12 @@ OlapScanOperatorFactory::OlapScanOperatorFactory(int32_t id, ScanNode* scan_node
         : ScanOperatorFactory(id, scan_node), _ctx_factory(std::move(ctx_factory)) {}
 
 Status OlapScanOperatorFactory::do_prepare(RuntimeState* state) {
+    auto olap_scan_node = dynamic_cast<OlapScanNode*>(_scan_node);
+    DCHECK(olap_scan_node != nullptr);
+    const TOlapScanNode& thrift_olap_scan_node = olap_scan_node->thrift_olap_scan_node();
+    const TupleDescriptor* tuple_desc = state->desc_tbl().get_tuple_descriptor(thrift_olap_scan_node.tuple_id);
+    DCHECK(tuple_desc != nullptr);
+    _ctx_factory->set_scan_table_id(tuple_desc->table_desc()->table_id());
```

Comment:
Could we make all the table ids ready when preparing `query_ctx`? Then we needn't lock any more.
But getting all the ids is a little bit trick and I don't know whether it proper, since we need some complex and low-maintenance processes:
1. traverse on `state->desc_tbl()::get_tuple_descs` 
2. check each `tuple_desc::table_desc` is non-null and instance of `OlapTableDescriptor`
3. get table id from `tuple_desc::table_desc`.

Reply:
> Could we make all the table ids ready when preparing `query_ctx`? Then we needn't lock any more. But getting all the ids is a little bit trick and I don't know whether it proper, since we need some complex and low-maintenance processes:
> 
> 1. traverse on `state->desc_tbl()::get_tuple_descs`
> 2. check each `tuple_desc::table_desc` is non-null and instance of `OlapTableDescriptor`
> 3. get table id from `tuple_desc::table_desc`.

this a little trick, after upgrading phmap, we can use parallel_flat_hash_map directly and the lock will be removed